### PR TITLE
Rewrite the parser

### LINKED ISSUE #62: Rewrite parser with megaparsec, separate passes.
Right now the parser is polymorphic (sub-parsers depend on the context they're called in). This makes the parser produce a semantically valid AST in a single pass, at the expense of error reporting (syntactically valid but semantically valid statements are rejected as parse errors).

To add insult to injury, the parser library is attoparsec, so parse errors are 99% of the time non informative.

A good solution would be to:

- use megaparsec, in order to get actually useful error messages on syntax errors
- make the parsing phase accept semantically invalid states, and forbid those in a separate pass, with appropriate error messages

It's quite a bit of work, but i've been working on a PoC a long time ago, with promising results. Working on this would also be a good opportunity to expose runtime parameter substitution in addition to compile-time slicing. Writing an error-reporting optimized megaparsec parser is extra work, but can be made incrementally. The baseline error reporting would be miles ahead of the current situation anyway.

### PR DESCRIPTION
**Pull request by divarvel**
Closes #62 

The main benefit is better error messages.

It also carries fixity fixes for operators, and support for runtime parameter substitution.

<details>
<summary>Before</summary>

```
λ> [block|fact($error, [$error, ["nested"]]); |]

<interactive>:3:8: error:
    • endOfInput
    • In the quasi-quotation:
        [block|fact($error, [$error, ["nested"]]); |]
```

</details>

<details>
<summary>After</summary>

```
λ> [block|fact($error, [$error, ["nested"]]); |]

<interactive>:8:8: error:
    • 1:6:
  |
1 | fact($error, [$error, ["nested"]]); 
  |      ^
Variables can't appear in a fact

1:15:
  |
1 | fact($error, [$error, ["nested"]]); 
  |               ^
Variables can't appear in a set

1:23:
  |
1 | fact($error, [$error, ["nested"]]); 
  |                       ^
Sets cannot be nested

    • In the quasi-quotation:
        [block|fact($error, [$error, ["nested"]]); |]
```

</details>

### CODE REVIEW

File: biscuit/test/Spec/Parser.hs [hunk 1]
```diff
@@ -26,31 +30,31 @@ pk4 = fromJust $ parsePublicKeyHex "c6864578bc03596d52878bd70025ec966c95c60727cb
...
-parseAuthorizer = parseOnly authorizerParser
+parseAuthorizer = (validationToEither . first show . substituteAuthorizer mempty mempty) <=< run authorizerParser
 
 parsePolicy :: Text -> Either String Policy
-parsePolicy = parseOnly policyParser
+parsePolicy = (validationToEither . first show . substitutePolicy mempty mempty) <=< run policyParser
 
 parseBlock :: Text -> Either String Block
-parseBlock = parseOnly blockParser
+parseBlock = (validationToEither . first show . substituteBlock mempty mempty) <=< run blockParser
```
Comment (ptitfred) [inline biscuit/test/Spec/Parser.hs:57]: This diff is screaming outloud: don't repeat yourself please 😬

File: biscuit/test/Spec/Executor.hs [hunk 2]
```diff
@@ -17,8 +16,9 @@ import           Auth.Biscuit.Datalog.Executor       (ExecutionError (..),
                                                       defaultLimits,
                                                       evaluateExpression)
 import           Auth.Biscuit.Datalog.Parser         (expressionParser, fact,
-                                                      rule)
+                                                      rule, run)
```
Comment (ptitfred) [inline biscuit/test/Spec/Executor.hs:19]: It doesn't show up in the diff thereafter ; is this import really useful?

**Comment by ptitfred** [comment]
Could you update the `extra-source-files:` section in the cabal to avoid breaking nixpkgs' biscuit-haskell on next release?

File: biscuit/test/Spec/Parser.hs (hunk 1 above)
Comment (divarvel) [reply biscuit/test/Spec/Parser.hs:57]: Ha, this part was written in anger. I'll factor the helper out

File: biscuit/test/Spec/Executor.hs (hunk 2 above)
Comment (divarvel) [reply biscuit/test/Spec/Executor.hs:19]: Oh, there is no -Wall on the tests, that's why it did not complain. I'll enable it and see what else explodes

**Comment by divarvel** [comment]
> Could you update the `extra-source-files:` section in the cabal to avoid breaking nixpkgs' biscuit-haskell on next release?

Ha, good catch

File: biscuit/src/Auth/Biscuit/Datalog/AST.hs [hunk 3]
```diff
@@ -744,7 +770,7 @@ renderRuleScope =
   let renderScopeElem = \case
         OnlyAuthority -> "authority"
         Previous      -> "previous"
-        BlockId bs    -> "ed25519/hex:" <> Hex.encodeBase16 (pkBytes bs)
+        BlockId bs    -> "ed25519/" <> Hex.encodeBase16 (pkBytes bs)
```
Comment (ptitfred) [inline biscuit/src/Auth/Biscuit/Datalog/AST.hs:773]: The spec has changed in the meantime?

File: biscuit/src/Auth/Biscuit/Datalog/Parser.hs [hunk 4]
```diff
@@ -1,448 +1,393 @@
...
-  parseBlockId :: Parser (BlockIdType evalCtx ctx)
-
-instance ScopeParser 'Repr 'Representation where
-  parseBlockId = publicKeyParser
+haskellVariableParser :: Parser Text
+haskellVariableParser = l $ do
+  _ <- chunk "${"
+  leadingUS <- optional $ C.char '_'
+  x <- if isJust leadingUS then C.letterChar else C.lowerChar
+  xs <- takeWhileP (Just "_, ', or any alphanumeric char") (\c -> c == '_' || c == '\'' || isAlphaNum c)
```
Comment (ptitfred) [inline biscuit/src/Auth/Biscuit/Datalog/Parser.hs:97]: Technically, you can have `'` in haskell variable names too.

File: biscuit/src/Auth/Biscuit/Datalog/Parser.hs [hunk 5]
```diff
@@ -1,448 +1,393 @@
...
+  _ <- C.char '}'
+  pure . maybe id T.cons leadingUS $ T.cons x xs
+
+setParser :: Parser (Set (Term' 'WithinSet 'InFact 'WithSlices))
+setParser = do
+  _ <- l $ C.char '['
+  ts <- sepBy (termParser (forbid VarInSet variableParser) (forbid NestedSet setParser)) (l $ C.char ',')
```
Comment (ptitfred) [inline biscuit/src/Auth/Biscuit/Datalog/Parser.hs:104]: No space allowed?

File: biscuit/src/Auth/Biscuit/Datalog/Parser.hs [hunk 6]
```diff
@@ -1,448 +1,393 @@
...
+  , Variable <$> parseVar
+  , TermSet <$> parseSet
+  , LBytes <$> (chunk "hex:" *> hexParser)
+  , LDate <$> try rfc3339DateParser
+  , LInteger <$> L.signed C.space L.decimal
+  , LString . T.pack <$> (C.char '"' *> manyTill L.charLiteral (C.char '"'))
+  , LBool <$> choice [ True <$ chunk "true"
+                     , False <$ chunk "false"
+                     ]
+  ]
```
Comment (ptitfred) [inline biscuit/src/Auth/Biscuit/Datalog/Parser.hs:130]: Have you tested it with funny cases like : variable prefixed by `hex` ?

1. Shouldn't `LBytes` be prioritized other `Variable`
2. I'm not sure if megaparsec backtracks by default in a choice, I used to do it explicitly with parsec. Is that needed here?

Comment (ptitfred) [reply biscuit/src/Auth/Biscuit/Datalog/Parser.hs:130]: > Shouldn't LBytes and LBool be prioritized other Variable?

I guess not, variables start with a dollar sign.

File: biscuit/src/Auth/Biscuit/Datalog/Parser.hs [hunk 7]
```diff
@@ -1,448 +1,393 @@
...
-unaryParens :: HasTopTermParsers 'InPredicate ctx => Parser (Expression' ctx)
-unaryParens = do
-  skipSpace
-  _ <- char '('
-  skipSpace
-  e <- expressionParser
-  skipSpace
-  _ <- char ')'
-  pure $ EUnary Parens e
+rfc3339DateParser :: Parser UTCTime
```
Comment (ptitfred) [inline biscuit/src/Auth/Biscuit/Datalog/Parser.hs:148]: Aouch

File: biscuit/src/Auth/Biscuit/Datalog/Parser.hs [hunk 8]
```diff
@@ -1,448 +1,393 @@
...
-  terms <- parens (commaList0 termParser)
-  pure Predicate{name,terms}
-
-ruleBodyParser :: (HasParsers 'InPredicate evalCtx ctx)
-               => Parser ([Predicate' 'InPredicate ctx], [Expression' ctx], Set.Set (RuleScope' evalCtx ctx))
+ruleParser :: Parser (Rule' 'Repr 'WithSlices)
+ruleParser = do
+  rhead <- l predicateParser
+  _ <- l $ chunk "<-"
+  (body, expressions, scope) <- ruleBodyParser
```
Comment (ptitfred) [inline biscuit/src/Auth/Biscuit/Datalog/Parser.hs:290]: Does it imply no space allowed around the left arrow ?

Comment (ptitfred) [reply biscuit/src/Auth/Biscuit/Datalog/Parser.hs:290]: Or that's the job of `l` ?

File: biscuit/src/Auth/Biscuit/Datalog/Parser.hs (hunk 6 above)
Comment (ptitfred) [reply biscuit/src/Auth/Biscuit/Datalog/Parser.hs:130]: > (does) megaparsec backtracks by default in a choice?

No it doesn't, it's the purpose of `try` you use elsewhere (like in parsec) and I guess it's not required here as there's no ambiguity in this subsyntax.

**Review by ptitfred — APPROVED** [review]
LGTM

File: biscuit/src/Auth/Biscuit/Datalog/AST.hs (hunk 3 above)
Comment (divarvel) [reply biscuit/src/Auth/Biscuit/Datalog/AST.hs:773]: yes https://github.com/biscuit-auth/biscuit/pull/117

File: biscuit/src/Auth/Biscuit/Datalog/Parser.hs (hunk 5 above)
Comment (divarvel) [reply biscuit/src/Auth/Biscuit/Datalog/Parser.hs:104]: `l` handles spaces: https://hackage.haskell.org/package/megaparsec-9.2.2/docs/Text-Megaparsec-Byte-Lexer.html#v:lexeme

File: biscuit/src/Auth/Biscuit/Datalog/Parser.hs (hunk 6 above)
Comment (divarvel) [reply biscuit/src/Auth/Biscuit/Datalog/Parser.hs:130]: antiquote, variable, term and bytes all start with `chunk`, which backtracks.

The only ambiguity is between integers and dates, hence the explicit `try`

File: biscuit/src/Auth/Biscuit/Datalog/Parser.hs (hunk 8 above)
Comment (divarvel) [reply biscuit/src/Auth/Biscuit/Datalog/Parser.hs:290]: that's `l`'s job. it handles whitespace (including comments)

**Pull request merged**